proposals/lip-0057.md

```diff
@@ -1910,6 +1916,46 @@ After the transactions in a block `b` are executed, the properties related to mi
 ```python
 def afterTransactionsExecute(b: Block) -> None:
     height = b.header.height
+    
+    updateProductivity(b.header)
+
+    # if not last block of round no further updates are needed.
+    if isEndOfRound(height) == False:
+        return
+
+    # Block b is an end-of-round block. Need to update snapshot and select validators for next round.
+    # This must be done after the properties related to missed blocks are updated.
+    
+    # No updates in validators and BFT weights are performed during bootstrap period.
+    if roundNumber(height+1) <= genesisDataStore.initRounds:
+        return
+
+    updateSnapshotSubstore(height)
+
+    # Calculate the active validators for next round based on the snapshot taken two rounds ago.
+    validatorList = getActiveValidators(snapshotStore(roundNumber(height)-2), roundNumber(height+1))
+
+    # Select standby validators if relevant.
+    if roundNumber(height+1) >= initRounds + NUMBER_ACTIVE_VALIDATORS:
+        standbyValidators = [validator for validator in snapshotStore(roundNumber(height)-2)
+                            if validator["address"] not in validatorList]
+
+        selectedStandbyValidators = getSelectedStandbyValidators(standbyValidators, height)
+        # Add selected standby validators to validators.
+        validatorList += selectedStandbyValidators
+
+    # If there are no eligible validators, validatorList is empty.
+    # In this case, the BFT parameters and validator list are not updated.
+    if validatorList is empty:
+        return
+
+    shuffleAndSetValidators(validatorList, height)
+
+
+
+def updateProductivity(b.header: BlockHeader) -> None:
```

Comment:
For clarity, could we put each of the functions in a separate code environment.

Reply:
Done.